src/api/mod.rs

```diff
@@ -18,68 +18,84 @@ use crate::api::{
     wallet::{get_balance, new_address, transfer},
     ws::ws_handler,
 };
-use anyhow::Result;
+use anyhow::{Context, Result};
 use api::routes;
 use axum::{
     extract::Extension,
     response::IntoResponse,
     routing::{delete, get, post},
     Router,
 };
-use axum_server::{tls_rustls::RustlsConfig, Handle};
+use axum_server::{
+    tls_rustls::{RustlsAcceptor, RustlsConfig},
+    Handle, Server,
+};
 use futures::{future::Shared, Future};
 use hyper::StatusCode;
 use log::{error, info};
 use std::{net::SocketAddr, sync::Arc, time::Duration};
 use tower_http::cors::CorsLayer;
 
-pub async fn start_rest_api(
-    listen_address: String,
-    certs_dir: String,
-    lightning_api: Arc<dyn LightningInterface + Send + Sync>,
-    wallet_api: Arc<dyn WalletInterface + Send + Sync>,
-    macaroon_auth: Arc<MacaroonAuth>,
-    quit_signal: Shared<impl Future<Output = ()>>,
-) -> Result<()> {
-    info!("Starting REST API");
-    let rustls_config = config(&certs_dir).await;
-    let cors = CorsLayer::permissive();
-    let handle = Handle::new();
-
-    let app = Router::new()
-        .route(routes::ROOT, get(root))
-        .route(routes::GET_INFO, get(get_info))
-        .route(routes::GET_BALANCE, get(get_balance))
-        .route(routes::LIST_CHANNELS, get(list_channels))
-        .route(routes::OPEN_CHANNEL, post(open_channel))
-        .route(routes::NEW_ADDR, get(new_address))
-        .route(routes::WITHDRAW, post(transfer))
-        .route(routes::LIST_PEERS, get(list_peers))
-        .route(routes::CONNECT_PEER, post(connect_peer))
-        .route(routes::DISCONNECT_PEER, delete(disconnect_peer))
-        .route(routes::WEBSOCKET, get(ws_handler))
-        .fallback(handler_404)
-        .layer(cors)
-        .layer(Extension(lightning_api))
-        .layer(Extension(wallet_api))
-        .layer(Extension(macaroon_auth));
+pub struct RestApi {
+    server: Server<RustlsAcceptor>,
+}
 
+pub async fn bind_api_server(listen_address: String, certs_dir: String) -> Result<RestApi> {
+    let rustls_config = config(&certs_dir)
+        .await
+        .context("failed to load tls configuration")?;
     let addr = listen_address.parse()?;
+    info!("Starting REST API on {addr}");
+    // FIXME: would be nice if failing to bind a socket would not actually panic
```

Comment:
I think its better not to put FIXMEs in the code (its distracting) and instead write an issue for them.

Reply:
Dropped the FIXME. I will stop adding FIXMEs in future.. In hindsight this one probably too much work to fix the crate to expose the api than gain from that.